docs/source/generics.rst

```diff
@@ -0,0 +1,984 @@
+:ref:`predefined protocols <predefined_protocols>` are generic, such as
+:py:class:`Iterable[T] <typing.Iterable>`, and you can define additional generic
+protocols. Generic protocols mostly follow the normal rules for generic classes.
+Example:
+
+.. code-block:: python
+
+   from typing import TypeVar
+   from typing_extensions import Protocol
+
+   T = TypeVar('T')
+
+   class Box(Protocol[T]):
+       content: T
+
+   def do_stuff(one: Box[str], other: Box[bytes]) -> None:
+       ...
+
+   class StringWrapper:
+       def __init__(self, content: str) -> None:
+           self.content = content
+
+   class BytesWrapper:
+       def __init__(self, content: bytes) -> None:
+           self.content = content
+
+   do_stuff(StringWrapper('one'), BytesWrapper(b'other'))  # OK
+
+   x: Box[float] = ...
+   y: Box[int] = ...
+   x = y  # Error -- Box is invariant
+
+Note that ``class ClassName(Protocol[T])`` is allowed as a shorthand for
+``class ClassName(Protocol, Generic[T])``, as per :pep:`PEP 544: Generic protocols <544#generic-protocols>`,
+
+The main difference between generic protocols and ordinary generic classes is
+that the declared variances of generic type variables in a protocol are checked
+against how they are used in the protocol definition.  The protocol in this
+example is rejected, since the type variable ``T`` is used covariantly as a
+return type, but the type variable is invariant:
+
+.. code-block:: python
+
+   from typing import Protocol, TypeVar
+
+   T = TypeVar('T')
+
+   class ReadOnlyBox(Protocol[T]):  # error: Invariant type variable "T" used in protocol where covariant one is expected
+       def content(self) -> T: ...
+
+This example correctly uses a covariant type variable:
+
+.. code-block:: python
+
+   from typing import Protocol, TypeVar
+
+   T_co = TypeVar('T_co', covariant=True)
+
+   class ReadOnlyBox(Protocol[T_co]):  # OK
+       def content(self) -> T_co: ...
+
+   ax: ReadOnlyBox[float] = ...
+   ay: ReadOnlyBox[int] = ...
+   ax = ay  # OK -- ReadOnlyBox is covariant
+
+See :ref:`variance-of-generics` for more about variance.
+
+Generic protocols can also be recursive. Example:
+
+.. code-block:: python
+
+   T = TypeVar('T')
+
+   class Linked(Protocol[T]):
+       val: T
+       def next(self) -> 'Linked[T]': ...
+
+   class L:
+       val: int
+       def next(self) -> 'L': ...
+
+   def last(seq: Linked[T]) -> T: ...
+
+   result = last(L())
+   reveal_type(result)  # Revealed type is "builtins.int"
+
+.. _generic-type-aliases:
+
+Generic type aliases
+********************
+
+Type aliases can be generic. In this case they can be used in two ways:
+Subscripted aliases are equivalent to original types with substituted type
+variables, so the number of type arguments must match the number of free type variables
+in the generic type alias. Unsubscripted aliases are treated as original types with free
+variables replaced with ``Any``. Examples (following :pep:`PEP 484: Type aliases
```

Comment:
Do type checkers always use `Any` regardless of the type parameter bound?

Reply:
Yes